File: module/ImaginaryCraft/src/main/java/architecture/imaginarycraft/common/particle/ItemStackParticle.java
package architecture.imaginarycraft.common.particle;

import architecture.goldenboughs_lib.util.LibUtil;
import architecture.imaginarycraft.init.IcParticleTypes;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import com.mojang.math.Axis;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.client.Camera;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.client.particle.Particle;
import net.minecraft.client.particle.ParticleProvider;
import net.minecraft.client.particle.ParticleRenderType;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.texture.OverlayTexture;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.util.Mth;
import net.minecraft.world.item.ItemDisplayContext;
import net.minecraft.world.item.ItemStack;

import java.util.ArrayList;
import java.util.List;

public class ItemStackParticle extends Particle {

    //region Constants & Static Fields

    public static final List<ItemStackParticle> PARTICLES = new ArrayList<>();

    //endregion

    //region Render Properties

    private ItemStack itemStack = ItemStack.EMPTY;
    private ItemDisplayContext itemDisplayContext = ItemDisplayContext.NONE;
    private double xRot = 0.0;
    private double xRotO = 0.0;
    private double yRot = 0.0;
    private double yRotO = 0.0;
    private double zRot = 0.0;
    private double zRotO = 0.0;
    private double scale = 1.0;
    private double scaleO = 1.0;

    //endregion

    //region Physics Properties

    private boolean enablePhysics = false;
    private double gravityScale = 1.0;
    private double bounce = 0.5;
    private double groundFriction = 0.8;
    private double xRotSpeed = 0.0;
    private double yRotSpeed = 0.0;
    private double zRotSpeed = 0.0;
    private boolean isFadingOut = false;
    private double scaleOnFadeStart = 1.0;

    //endregion

    //region Constructors

    public ItemStackParticle(ClientLevel level, double x, double y, double z, double xSpeed, double ySpeed, double zSpeed) {
        super(level, x, y, z, xSpeed, ySpeed, zSpeed);
        PARTICLES.add(this);
    }

    public ItemStackParticle(ClientLevel level, double x, double y, double z) {
        super(level, x, y, z);
        PARTICLES.add(this);
    }

    //endregion

    //region Getters & Setters

    public ItemStack getItemStack() {
        return itemStack;
    }

    public void setItemStack(ItemStack itemStack) {
        this.itemStack = itemStack;
    }

    public ItemDisplayContext getItemDisplayContext() {
        return itemDisplayContext;
    }

    public void setItemDisplayContext(ItemDisplayContext itemDisplayContext) {
        this.itemDisplayContext = itemDisplayContext;
    }

    public double getXRot() {
        return xRot;
    }

    public void setXRot(double xRot) {
        this.xRot = xRot;
    }

    public double getXRotO() {
        return xRotO;
    }

    public void setXRotO(double xRotO) {
        this.xRotO = xRotO;
    }

    public double getYRot() {
        return yRot;
    }

    public void setYRot(double yRot) {
        this.yRot = yRot;
    }

    public double getYRotO() {
        return yRotO;
    }

    public void setYRotO(double yRotO) {
        this.yRotO = yRotO;
    }

    public double getZRot() {
        return zRot;
    }

    public void setZRot(double zRot) {
        this.zRot = zRot;
    }

    public double getZRotO() {
        return zRotO;
    }

    public void setZRotO(double zRotO) {
        this.zRotO = zRotO;
    }

    public double getScale() {
        return scale;
    }

    public void setScale(double scale) {
        this.scale = scale;
    }

    public double getScaleO() {
        return scaleO;
    }

    public void setScaleO(double scaleO) {
        this.scaleO = scaleO;
    }

    public boolean isEnablePhysics() {
        return enablePhysics;
    }

    public void setEnablePhysics(boolean enablePhysics) {
        this.enablePhysics = enablePhysics;
    }

    public double getGravityScale() {
        return gravityScale;
    }

    public void setGravityScale(double gravityScale) {
        this.gravityScale = gravityScale;
    }

    public double getBounce() {
        return bounce;
    }

    public void setBounce(double bounce) {
        this.bounce = bounce;
    }

    public double getGroundFriction() {
        return groundFriction;
    }

    public void setGroundFriction(double groundFriction) {
        this.groundFriction = groundFriction;
    }

    public double getXRotSpeed() {
        return xRotSpeed;
    }

    public void setXRotSpeed(double xRotSpeed) {
        this.xRotSpeed = xRotSpeed;
    }

    public double getYRotSpeed() {
        return yRotSpeed;
    }

    public void setYRotSpeed(double yRotSpeed) {
        this.yRotSpeed = yRotSpeed;
    }

    public double getZRotSpeed() {
        return zRotSpeed;
    }

    public void setZRotSpeed(double zRotSpeed) {
        this.zRotSpeed = zRotSpeed;
    }

    public boolean isFadingOut() {
        return isFadingOut;
    }

    public void setFadingOut(boolean fadingOut) {
        isFadingOut = fadingOut;
    }

    //endregion

    //region Interpolated Query Methods

    public double getYRot(float partialTicks) {
        return Mth.lerp((double) partialTicks, yRotO, yRot);
    }

    public double getXRot(float partialTicks) {
        return Mth.lerp((double) partialTicks, xRotO, xRot);
    }

    public double getZRot(float partialTicks) {
        return Mth.lerp((double) partialTicks, zRotO, zRot);
    }

    public double getY(float partialTicks) {
        return Mth.lerp((double) partialTicks, yo, y);
    }

    public double getX(float partialTicks) {
        return Mth.lerp((double) partialTicks, zo, x);
    }

    public double getZ(float partialTicks) {
        return Mth.lerp((double) partialTicks, zo, z);
    }

    public double getScale(float partialTicks) {
        return Mth.lerp((double) partialTicks, scaleO, scale);
    }

    //endregion

    //region Render

    @Override
    public void render(VertexConsumer buffer, Camera camera, float partialTicks) {
        // Empty implementation
    }

    public void render(PoseStack poseStack, Camera camera, float partialTicks, MultiBufferSource bufferSource) {
        poseStack.pushPose();
        int combinedLight = getLightColor(partialTicks);
        int combinedOverlay = OverlayTexture.NO_OVERLAY;
        var itemRenderer = Minecraft.getInstance().getItemRenderer();

        poseStack.translate((float) getX(partialTicks), (float) getY(partialTicks), (float) getZ(partialTicks));
        poseStack.mulPose(Axis.XP.rotationDegrees((float) getXRot(partialTicks)));
        poseStack.mulPose(Axis.YP.rotationDegrees((float) getYRot(partialTicks)));
        poseStack.mulPose(Axis.ZP.rotationDegrees((float) getZRot(partialTicks)));
        float s = (float) getScale(partialTicks);
        poseStack.scale(s, s, s);
        itemRenderer.renderStatic(itemStack, itemDisplayContext, combinedLight, combinedOverlay, poseStack, bufferSource, level, 0);
        poseStack.popPose();
    }

    @Override
    public ParticleRenderType getRenderType() {
        return ParticleRenderType.NO_RENDER;
    }

    //endregion

    //region Tick

    @Override
    public void tick() {
        xRotO = xRot;
        yRotO = yRot;
        zRotO = zRot;
        scaleO = scale;

        if (enablePhysics) {
            tickPhysics();
        } else {
            super.tick();
        }
    }

    private void tickPhysics() {
        xo = x;
        yo = y;
        zo = z;

        if (this.age++ >= this.lifetime) {
            this.remove();
            return;
        }

        double prevX = x;
        double prevY = y;
        double prevZ = z;

        this.yd -= gravityScale * (double) this.gravity;
        this.move(this.xd, this.yd, this.zd);

        if (this.speedUpWhenYMotionIsBlocked && this.y == this.yo) {
            this.xd *= 1.1;
            this.zd *= 1.1;
        }

        this.xd *= (double) this.friction;
        this.yd *= (double) this.friction;
        this.zd *= (double) this.friction;
        if (this.onGround) {
            this.xd *= 0.7;
            this.zd *= 0.7;
        }

        xRot += xRotSpeed;
        yRot += yRotSpeed;
        zRot += zRotSpeed;

        boolean collidedX = x == prevX && xd != 0.0;
        boolean collidedY = y == prevY && yd != 0.0;
        boolean collidedZ = z == prevZ && zd != 0.0;

        if (collidedX || collidedY || collidedZ) {
            onCollision(collidedX, collidedY, collidedZ);
        }

        if (collidedY) yd = -yd * bounce;
        if (collidedX) xd = -xd * bounce;
        if (collidedZ) zd = -zd * bounce;

        if (isFadingOut) {
            double progress = (double) age / (double) lifetime;
            scale = scaleOnFadeStart * (1.0 - progress);
        }
    }

    //endregion

    //region Collision

    public void onCollision(boolean collidedX, boolean collidedY, boolean collidedZ) {
        if (!isFadingOut) {
            isFadingOut = true;
            scaleOnFadeStart = scale;
        }
    }

    //endregion

    //region Lifecycle

    @Override
    public void remove() {
        super.remove();
        PARTICLES.remove(this);
    }

    //endregion

    //region Inner Types

    public static class Provider implements ParticleProvider<Options> {
        @Override
        public Particle createParticle(Options type, ClientLevel level, double x, double y, double z, double xSpeed, double ySpeed, double zSpeed) {
            ItemStackParticle particle = new ItemStackParticle(level, x, y, z, xSpeed, ySpeed, zSpeed);
            particle.itemStack = type.itemStack();
            particle.itemDisplayContext = type.itemDisplayContext();
            particle.xRot = type.xRot();
            particle.xRotO = type.xRot();
            particle.yRot = type.yRot();
            particle.yRotO = type.yRot();
            particle.zRot = type.zRot();
            particle.zRotO = type.zRot();
            particle.enablePhysics = type.enablePhysics();
            particle.gravityScale = type.gravityScale();
            particle.bounce = type.bounce();
            particle.groundFriction = type.groundFriction();
            particle.xRotSpeed = type.xRotSpeed();
            particle.yRotSpeed = type.yRotSpeed();
            particle.zRotSpeed = type.zRotSpeed();
            particle.lifetime = type.lifetime();
            return particle;
        }
    }

    public record Options(
            ItemStack itemStack,
            ItemDisplayContext itemDisplayContext,
            double xRot,
            double yRot,
            double zRot,
            boolean enablePhysics,
            double gravityScale,
            double bounce,
            double groundFriction,
            double xRotSpeed,
            double yRotSpeed,
            double zRotSpeed,
            int lifetime
    ) implements ParticleOptions {

        public static final MapCodec<Options> CODEC = RecordCodecBuilder.mapCodec(instance ->
                instance.group(
                        ItemStack.CODEC.fieldOf("itemStack").forGetter(Options::itemStack),
                        ItemDisplayContext.CODEC.fieldOf("itemDisplayContext").forGetter(Options::itemDisplayContext),
                        Codec.DOUBLE.optionalFieldOf("xRot", 0.0).forGetter(Options::xRot),
                        Codec.DOUBLE.optionalFieldOf("yRot", 0.0).forGetter(Options::yRot),
                        Codec.DOUBLE.optionalFieldOf("zRot", 0.0).forGetter(Options::zRot),
                        Codec.BOOL.optionalFieldOf("enablePhysics", false).forGetter(Options::enablePhysics),
                        Codec.DOUBLE.optionalFieldOf("gravityScale", 1.0).forGetter(Options::gravityScale),
                        Codec.DOUBLE.optionalFieldOf("bounce", 0.5).forGetter(Options::bounce),
                        Codec.DOUBLE.optionalFieldOf("groundFriction", 0.8).forGetter(Options::groundFriction),
                        Codec.DOUBLE.optionalFieldOf("xRotSpeed", 0.0).forGetter(Options::xRotSpeed),
                        Codec.DOUBLE.optionalFieldOf("yRotSpeed", 0.0).forGetter(Options::yRotSpeed),
                        Codec.DOUBLE.optionalFieldOf("zRotSpeed", 0.0).forGetter(Options::zRotSpeed),
                        Codec.INT.optionalFieldOf("lifetime", 100).forGetter(Options::lifetime)
                ).apply(instance, Options::new)
        );

        public static final StreamCodec<RegistryFriendlyByteBuf, Options> STREAM_CODEC = StreamCodec.of(
                (buf, p) -> {
                    ItemStack.STREAM_CODEC.encode(buf, p.itemStack());
                    LibUtil.ITEM_DISPLAY_CONTEXT_STREAM_CODEC.encode(buf, p.itemDisplayContext());
                    ByteBufCodecs.DOUBLE.encode(buf, p.xRot());
                    ByteBufCodecs.DOUBLE.encode(buf, p.yRot());
                    ByteBufCodecs.DOUBLE.encode(buf, p.zRot());
                    ByteBufCodecs.BOOL.encode(buf, p.enablePhysics());
                    ByteBufCodecs.DOUBLE.encode(buf, p.gravityScale());
                    ByteBufCodecs.DOUBLE.encode(buf, p.bounce());
                    ByteBufCodecs.DOUBLE.encode(buf, p.groundFriction());
                    ByteBufCodecs.DOUBLE.encode(buf, p.xRotSpeed());
                    ByteBufCodecs.DOUBLE.encode(buf, p.yRotSpeed());
                    ByteBufCodecs.DOUBLE.encode(buf, p.zRotSpeed());
                    ByteBufCodecs.INT.encode(buf, p.lifetime());
                },
                buf -> new Options(
                        ItemStack.STREAM_CODEC.decode(buf),
                        LibUtil.ITEM_DISPLAY_CONTEXT_STREAM_CODEC.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.BOOL.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.DOUBLE.decode(buf),
                        ByteBufCodecs.INT.decode(buf)
                )
        );

        /**
         * 使用默认值创建 Options。默认 lifetime 为 100 ticks（5 秒）。
         */
        public static Options of(ItemStack itemStack) {
            return new Options(itemStack, ItemDisplayContext.NONE, 0.0, 0.0, 0.0, false, 1.0, 0.5, 0.8, 0.0, 0.0, 0.0, 100);
        }

        @Override
        public ParticleType<?> getType() {
            return IcParticleTypes.ITEM_STACK.get();
        }
    }

    //endregion
}

File: module/ImaginaryCraft/src/main/kotlin/architecture/imaginarycraft/common/particle/ItemStackParticle.kt
package architecture.imaginarycraft.common.particle

import architecture.goldenboughs_lib.api.AllOpe
import architecture.goldenboughs_lib.util.LibUtil.ITEM_DISPLAY_CONTEXT_STREAM_CODEC
import architecture.imaginarycraft.init.IcParticleTypes
import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexConsumer
import com.mojang.math.Axis
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.client.Camera
import net.minecraft.client.Minecraft
import net.minecraft.client.multiplayer.ClientLevel
import net.minecraft.client.particle.Particle
import net.minecraft.client.particle.ParticleProvider
import net.minecraft.client.particle.ParticleRenderType
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.core.particles.ParticleOptions
import net.minecraft.core.particles.ParticleType
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.util.Mth
import net.minecraft.world.item.ItemDisplayContext
import net.minecraft.world.item.ItemStack

// TODO 修复粒子问题
@AllOpe
class ItemStackParticle : Particle {

	//region 静态列表

	companion object {
		@JvmStatic
		val PARTICLES = mutableListOf<ItemStackParticle>()
	}

	//endregion

	//region 渲染属性

	var itemStack: ItemStack = ItemStack.EMPTY
	var itemDisplayContext: ItemDisplayContext = ItemDisplayContext.NONE
	var xRot: Double = 0.0
	var xRotO: Double = 0.0
	var yRot: Double = 0.0
	var yRotO: Double = 0.0
	var zRot: Double = 0.0
	var zRotO: Double = 0.0

	/** 渲染缩放比例（1.0=原始大小，可用于淡出缩小等效果） */
	var scale: Double = 1.0

	/** 上一帧的渲染缩放比例，用于插值 */
	var scaleO: Double = 1.0

	//endregion

	//region 物理属性

	/** 是否启用物理效果（重力、碰撞、弹力、摩擦、角速度等） */
	var enablePhysics: Boolean = false

	/** 重力加速度倍数，默认 1.0 为标准重力 */
	var gravityScale: Double = 1.0

	/** 弹力系数，0=无弹力，1=完全弹性，默认 0.5 */
	var bounce: Double = 0.5

	/** 地面摩擦力，默认 0.8（每次 tick 速度乘以该值） */
	var groundFriction: Double = 0.8

	/** X 轴角速度（度/tick） */
	var xRotSpeed: Double = 0.0

	/** Y 轴角速度（度/tick） */
	var yRotSpeed: Double = 0.0

	/** Z 轴角速度（度/tick） */
	var zRotSpeed: Double = 0.0

	/** 是否已进入碰撞淡出状态 */
	var isFadingOut: Boolean = false

	/** 进入淡出时的缩放值 */
	private var scaleOnFadeStart: Double = 1.0

	//endregion

	//region 构造器

	constructor(
		level: ClientLevel,
		x: Double,
		y: Double,
		z: Double,
		xSpeed: Double,
		ySpeed: Double,
		zSpeed: Double
	) : super(level, x, y, z, xSpeed, ySpeed, zSpeed) {
		PARTICLES.add(this)
	}

	constructor(
		level: ClientLevel,
		x: Double,
		y: Double,
		z: Double
	) : super(level, x, y, z) {
		PARTICLES.add(this)
	}

	//endregion

	//region 插值查询

	fun getYRot(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), yRotO, yRot)

	fun getXRot(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), xRotO, xRot)

	fun getZRot(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), zRotO, zRot)

	fun getY(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), yo, y)

	fun getX(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), zo, x)

	fun getZ(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), zo, z)

	fun getScale(partialTicks: Float): Double =
		Mth.lerp(partialTicks.toDouble(), scaleO, scale)

	//endregion

	//region 渲染

	override fun render(
		buffer: VertexConsumer,
		camera: Camera,
		partialTicks: Float
	) {
	}

	fun render(
		poseStack: PoseStack,
		camera: Camera,
		partialTicks: Float,
		bufferSource: MultiBufferSource
	) {
		poseStack.pushPose()
		val combinedLight = getLightColor(partialTicks)
		val combinedOverlay = OverlayTexture.NO_OVERLAY
		val itemRenderer = Minecraft.getInstance().itemRenderer

		poseStack.translate(
			getX(partialTicks).toFloat(),
			getY(partialTicks).toFloat(),
			getZ(partialTicks).toFloat()
		)
		poseStack.mulPose(Axis.XP.rotationDegrees(getXRot(partialTicks).toFloat()))
		poseStack.mulPose(Axis.YP.rotationDegrees(getYRot(partialTicks).toFloat()))
		poseStack.mulPose(Axis.ZP.rotationDegrees(getZRot(partialTicks).toFloat()))
		val s = getScale(partialTicks).toFloat()
		poseStack.scale(s, s, s)
		itemRenderer.renderStatic(
			itemStack,
			itemDisplayContext,
			combinedLight,
			combinedOverlay,
			poseStack,
			bufferSource,
			level,
			0
		)
		poseStack.popPose()
	}

	override fun getRenderType(): ParticleRenderType = ParticleRenderType.NO_RENDER

	//endregion

	//region Tick

	override fun tick() {
		xRotO = xRot
		yRotO = yRot
		zRotO = zRot
		scaleO = scale

		if (enablePhysics) {
			tickPhysics()
		} else {
			super.tick()
		}
	}

	/**
	 * 物理 tick 逻辑：重力、角速度、碰撞、反弹、摩擦、淡出缩小。
	 */
	private fun tickPhysics() {
		xo = x
		yo = y
		zo = z

		if (this.age++ >= this.lifetime) {
			this.remove()
			return
		}

		val prevX = x
		val prevY = y
		val prevZ = z

		this.yd -= gravityScale * this.gravity.toDouble()
		this.move(this.xd, this.yd, this.zd)
		if (this.speedUpWhenYMotionIsBlocked && this.y == this.yo) {
			this.xd *= 1.1
			this.zd *= 1.1
		}

		this.xd *= this.friction.toDouble()
		this.yd *= this.friction.toDouble()
		this.zd *= this.friction.toDouble()
		if (this.onGround) {
			this.xd *= 0.7
			this.zd *= 0.7
		}

		xRot += xRotSpeed
		yRot += yRotSpeed
		zRot += zRotSpeed

		val collidedX = x == prevX && xd != 0.0
		val collidedY = y == prevY && yd != 0.0
		val collidedZ = z == prevZ && zd != 0.0

		if (collidedX || collidedY || collidedZ) {
			onCollision(collidedX, collidedY, collidedZ)
		}

		if (collidedY) yd = -yd * bounce
		if (collidedX) xd = -xd * bounce
		if (collidedZ) zd = -zd * bounce

		if (isFadingOut) {
			val progress = age.toDouble() / lifetime.toDouble()
			scale = scaleOnFadeStart * (1.0 - progress)
		}
	}

	//endregion

	//region 碰撞回调

	/**
	 * 粒子碰撞时调用（仅在 [enablePhysics] 为 true 时触发）。
	 * 默认行为：触发缩小淡出（不立即移除）。
	 * 子类可重写此方法实现自定义碰撞行为。
	 */
	fun onCollision(collidedX: Boolean, collidedY: Boolean, collidedZ: Boolean) {
		if (!isFadingOut) {
			isFadingOut = true
			scaleOnFadeStart = scale
		}
	}

	//endregion

	//region 生命周期

	override fun remove() {
		super.remove()
		PARTICLES.remove(this)
	}

	//endregion

	//region Provider

	class Provider : ParticleProvider<Options> {
		override fun createParticle(
			type: Options,
			level: ClientLevel,
			x: Double,
			y: Double,
			z: Double,
			xSpeed: Double,
			ySpeed: Double,
			zSpeed: Double
		): Particle {
			return ItemStackParticle(level, x, y, z, xSpeed, ySpeed, zSpeed).apply {
				itemStack = type.itemStack
				itemDisplayContext = type.itemDisplayContext
				xRot = type.xRot
				xRotO = type.xRot
				yRot = type.yRot
				yRotO = type.yRot
				zRot = type.zRot
				zRotO = type.zRot
				enablePhysics = type.enablePhysics
				gravityScale = type.gravityScale
				bounce = type.bounce
				groundFriction = type.groundFriction
				xRotSpeed = type.xRotSpeed
				yRotSpeed = type.yRotSpeed
				zRotSpeed = type.zRotSpeed
				lifetime = type.lifetime
			}
		}
	}

	//endregion

	//region Options

	/** Options 不包含 scale/scaleO，它们是运行时状态，由粒子本地管理。 */
	@JvmRecord
	data class Options(
		val itemStack: ItemStack,
		val itemDisplayContext: ItemDisplayContext = ItemDisplayContext.NONE,
		val xRot: Double = 0.0,
		val yRot: Double = 0.0,
		val zRot: Double = 0.0,
		val enablePhysics: Boolean = false,
		val gravityScale: Double = 1.0,
		val bounce: Double = 0.5,
		val groundFriction: Double = 0.8,
		val xRotSpeed: Double = 0.0,
		val yRotSpeed: Double = 0.0,
		val zRotSpeed: Double = 0.0,
		val lifetime: Int = 20 * 5
	) : ParticleOptions {
		companion object {

			@JvmField
			val CODEC: MapCodec<Options> = RecordCodecBuilder.mapCodec { instance ->
				instance.group(
					ItemStack.CODEC.fieldOf("itemStack").forGetter { o: Options -> o.itemStack },
					ItemDisplayContext.CODEC.fieldOf("itemDisplayContext").forGetter { o: Options -> o.itemDisplayContext },
					Codec.DOUBLE.optionalFieldOf("xRot", 0.0).forGetter { o: Options -> o.xRot },
					Codec.DOUBLE.optionalFieldOf("yRot", 0.0).forGetter { o: Options -> o.yRot },
					Codec.DOUBLE.optionalFieldOf("zRot", 0.0).forGetter { o: Options -> o.zRot },
					Codec.BOOL.optionalFieldOf("enablePhysics", false).forGetter { o: Options -> o.enablePhysics },
					Codec.DOUBLE.optionalFieldOf("gravityScale", 1.0).forGetter { o: Options -> o.gravityScale },
					Codec.DOUBLE.optionalFieldOf("bounce", 0.5).forGetter { o: Options -> o.bounce },
					Codec.DOUBLE.optionalFieldOf("groundFriction", 0.8).forGetter { o: Options -> o.groundFriction },
					Codec.DOUBLE.optionalFieldOf("xRotSpeed", 0.0).forGetter { o: Options -> o.xRotSpeed },
					Codec.DOUBLE.optionalFieldOf("yRotSpeed", 0.0).forGetter { o: Options -> o.yRotSpeed },
					Codec.DOUBLE.optionalFieldOf("zRotSpeed", 0.0).forGetter { o: Options -> o.zRotSpeed },
					Codec.INT.optionalFieldOf("lifetime", 20 * 5).forGetter { o: Options -> o.lifetime }
				).apply(instance, ::Options)
			}

			@JvmField
			val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf, Options> = StreamCodec.of(
				{ buf, p ->
					ItemStack.STREAM_CODEC.encode(buf, p.itemStack)
					ITEM_DISPLAY_CONTEXT_STREAM_CODEC.encode(buf, p.itemDisplayContext)
					ByteBufCodecs.DOUBLE.encode(buf, p.xRot)
					ByteBufCodecs.DOUBLE.encode(buf, p.yRot)
					ByteBufCodecs.DOUBLE.encode(buf, p.zRot)
					ByteBufCodecs.BOOL.encode(buf, p.enablePhysics)
					ByteBufCodecs.DOUBLE.encode(buf, p.gravityScale)
					ByteBufCodecs.DOUBLE.encode(buf, p.bounce)
					ByteBufCodecs.DOUBLE.encode(buf, p.groundFriction)
					ByteBufCodecs.DOUBLE.encode(buf, p.xRotSpeed)
					ByteBufCodecs.DOUBLE.encode(buf, p.yRotSpeed)
					ByteBufCodecs.DOUBLE.encode(buf, p.zRotSpeed)
					ByteBufCodecs.INT.encode(buf, p.lifetime)
				}, { buf ->
					Options(
						ItemStack.STREAM_CODEC.decode(buf),
						ITEM_DISPLAY_CONTEXT_STREAM_CODEC.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.BOOL.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.DOUBLE.decode(buf),
						ByteBufCodecs.INT.decode(buf)
					)
				}
			)
		}

		override fun getType(): ParticleType<*> = IcParticleTypes.ITEM_STACK.get()
	}

	//endregion
}
